File: backend/app/routers/predict.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from app.auth import get_current_user
from app.services.ml_service import ml_service
from app.services.privacy import redact_predict_response

router = APIRouter(prefix="/predict", tags=["predict"])


class SensorEvent(BaseModel):
    time: str
    sensor: str
    state: str


class PredictRequest(BaseModel):
    date: str
    events: list[SensorEvent]


@router.post("")
def predict_day(request: PredictRequest, user: dict = Depends(get_current_user)):
    events = [e.model_dump() for e in request.events]
    results = ml_service.predict_day(events)
    response = {
        "date": request.date,
        "rooms": results,
        "total_events": len(request.events),
    }
    return redact_predict_response(response, user["role"])

File: frontend/src/model/predict.ts
import { z } from "zod/v4";
import { RoomInfoSchema } from "./day";

export const SensorEventSchema = z.object({
    time: z.string(),
    sensor: z.string(),
    state: z.string(),
});

export type SensorEvent = z.infer<typeof SensorEventSchema>;

export const PredictRequestSchema = z.object({
    date: z.string(),
    events: z.array(SensorEventSchema),
});

export type PredictRequest = z.infer<typeof PredictRequestSchema>;

export const PredictResponseSchema = z.object({
    date: z.string(),
    rooms: z.record(z.string(), RoomInfoSchema),
    total_events: z.number(),
    redacted_rooms: z.array(z.string()).optional().default([]),
});

export type PredictResponse = z.infer<typeof PredictResponseSchema>;
